lib/elixir/lib/string.ex

```diff
@@ -1897,17 +1897,29 @@ defmodule String do
       true
 
   """
-  @spec starts_with?(t, t | [t]) :: boolean
-  def starts_with?(string, []) when is_binary(string) do
-    false
+  @spec starts_with?(t, pattern) :: boolean
+  def starts_with?(string, prefix) when is_binary(string) and is_binary(prefix) do
+    starts_with_string?(string, byte_size(string), prefix)
   end
 
   def starts_with?(string, prefix) when is_binary(string) and is_list(prefix) do
-    "" in prefix or Kernel.match?({0, _}, :binary.match(string, prefix))
+    string_size = byte_size(string)
+    Enum.any?(prefix, &starts_with_string?(string, string_size, &1))
   end
 
   def starts_with?(string, prefix) when is_binary(string) do
```

Comment:
Looking at the doc and the specs, the first two `starts_with?/2` should catch all the possible valid combinations (a string and a list of strings). In which cases this function will be executed? 🤔

Reply:
Nevermind. I saw `:binary.cp()` in the `pattern` typespec.

Reply:
Yes, it wasn't previously documented with the typespecs, but it's allowed.